### ISSUE TITLE
dtype of methods in tensorflow.contrib.distributions does not depend on self.dtype

### ISSUE THREAD

**Issue by dustinvtran** [tensorflow/tensorflow#8593]
In master (f4a0c2c) (and versions such as r1.0), the docstring in `log_prob()` claims to return a tensor which has the same dtype as the distributions' dtype. This is not true for the built-in distributions. For example:
```python
>>> from tensorflow.contrib import distributions as ds
>>>>
>>> x = ds.Bernoulli(p=0.5)
>>> x.dtype
tf.int32
>>> x.log_prob(1)
<tf.Tensor 'Bernoulli_1/log_prob/Neg:0' shape=() dtype=float32>
```

This inconsistency also occurs for many methods beyond `log_prob()`.

**Comment by skye** [comment]
@ebrevdo @jvdillon @langmore is this just a problem in documentation?

**Comment by ebrevdo** [comment]
Perhaps it is a documentation problem.  the dtype of a distribution is the
dtype of the *samples it emits *(the dtype of dist.sample(...)).  log_prob
is always going to depend on the dtypes of the logits / probabilities /
whatever you passed to the initializer of the distribution.

**Comment by dustinvtran** [comment]
In Bernoulli, sample() casts the output to `self.dtype` before returning it. So it has the same dtype as the distribution's.
```python
>>> from tensorflow.contrib import distributions as ds
>>>>
>>> x = ds.Bernoulli(p=0.5)
>>> x.dtype
tf.int32
>>> x.log_prob(1)
<tf.Tensor 'Bernoulli_1/log_prob/Neg:0' shape=() dtype=float32>
>>> x.sample()
<tf.Tensor 'Bernoulli/sample/Reshape:0' shape=() dtype=int32>
```
Not sure about the others. Are there cases where the dtype returned by sample() is different from the distribution's dtype?

**Comment by ebrevdo** [comment]
I don't know of any distributions where sample().dtype != dist.dtype.

Most other methods should not be expected to match the dtype, because they
include computations on the probability and the domain of the
distribution.  All statistics, log_prob, etc, require computations that mix
the domain values (of type dist.dtype) of the probabilities (usually one of
tf.float{16,32,64}).
